File: packages/next/src/middleware/ip-audit-hook.ts
import { createIpAuditEvent, type AuthNinjaConfig } from "@auth-ninja/core";
import { persistAuditEvent } from "../audit/persist.js";
import type { AuthDb } from "../db/client.js";
import type { RequestMeta } from "../http/request-meta.js";

export type IpAuditHookResult = {
  blocked: boolean;
  reason?: "allowlist_violation";
};

function isIpAllowed(ipAddress: string, allowlist: string[] | undefined): boolean {
  if (!allowlist || allowlist.length === 0) {
    return true;
  }
  return allowlist.includes(ipAddress);
}

/**
 * Enforce optional IP allowlist and persist IP audit events when policy triggers.
 * Returns `{ blocked: true }` when the client IP is not on the allowlist.
 */
export async function runIpAuditHook(
  config: AuthNinjaConfig,
  db: AuthDb,
  meta: RequestMeta,
): Promise<IpAuditHookResult> {
  if (!config.ipAuditEnabled) {
    return { blocked: false };
  }

  if (!isIpAllowed(meta.ipAddress, config.ipAllowlist)) {
    const audit = createIpAuditEvent({
      ipAddress: meta.ipAddress,
      userAgent: meta.userAgent,
      reason: "allowlist_violation",
    });
    await persistAuditEvent(db, audit);
    return { blocked: true, reason: "allowlist_violation" };
  }

  return { blocked: false };
}

/** Record a suspicious-activity IP audit event (e.g. after rate-limit breach). */
export async function recordSuspiciousIpAudit(
  config: AuthNinjaConfig,
  db: AuthDb,
  meta: RequestMeta,
): Promise<void> {
  if (!config.ipAuditEnabled) {
    return;
  }

  const audit = createIpAuditEvent({
    ipAddress: meta.ipAddress,
    userAgent: meta.userAgent,
    reason: "suspicious_activity",
  });
  await persistAuditEvent(db, audit);
}

File: adapters/dotnet/src/AuthNinja.AspNetCore/Middleware/IpAuditHook.cs
using AuthNinja.AspNetCore.Auth.Services;
using AuthNinja.AspNetCore.Http;
using Microsoft.AspNetCore.Http;

namespace AuthNinja.AspNetCore.Middleware;

internal static class IpAuditHook
{
    public static async Task<IpAuditHookResult> RunAsync(
        AuthNinjaOptions config,
        AuditService audit,
        HttpContext context,
        CancellationToken cancellationToken = default)
    {
        if (!config.IpAuditEnabled)
        {
            return new IpAuditHookResult(Blocked: false);
        }

        var ipAddress = RequestMeta.GetIpAddress(context);
        if (IsIpAllowed(ipAddress, config.IpAllowlist))
        {
            return new IpAuditHookResult(Blocked: false);
        }

        await audit.PersistIpAuditAsync(
            "allowlist_violation",
            ipAddress,
            RequestMeta.GetUserAgent(context),
            DateTimeOffset.UtcNow,
            cancellationToken);

        return new IpAuditHookResult(Blocked: true, Reason: "allowlist_violation");
    }

    public static async Task RecordSuspiciousActivityAsync(
        AuthNinjaOptions config,
        AuditService audit,
        HttpContext context,
        CancellationToken cancellationToken = default)
    {
        if (!config.IpAuditEnabled)
        {
            return;
        }

        await audit.PersistIpAuditAsync(
            "suspicious_activity",
            RequestMeta.GetIpAddress(context),
            RequestMeta.GetUserAgent(context),
            DateTimeOffset.UtcNow,
            cancellationToken);
    }

    private static bool IsIpAllowed(string ipAddress, string[]? allowlist) =>
        allowlist is not { Length: > 0 } || allowlist.Contains(ipAddress);
}

internal readonly record struct IpAuditHookResult(bool Blocked, string? Reason = null);
